src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs

```diff
@@ -602,35 +501,88 @@ public async Task<IRule> GetRuleAsync(IDependency dependency, IProjectCatalogSna
                 itemType: dependency.SchemaItemType,
                 itemName: itemSpec);
 
-            IRule rule = null;
-            if (schema != null)
+            if (schema == null)
             {
-                if (dependency.Resolved)
+                // Since we have no browse object, we still need to create *something* so
+                // that standard property pages can pop up.
+                Rule emptyRule = RuleExtensions.SynthesizeEmptyRule(context.ItemType);
+                return configuredProjectExports.PropertyPagesDataModelProvider.GetRule(
+                    emptyRule,
+                    context.File,
+                    context.ItemType,
+                    context.ItemName);
+            }
+
+            if (dependency.Resolved)
+            {
+                return configuredProjectExports.RuleFactory.CreateResolvedReferencePageRule(
+                    schema,
+                    context,
+                    dependency.Name,
+                    dependency.Properties);
+            }
+
+            return browseObjectsCatalog.BindToContext(schema.Name, context);
+
+            async Task<IImmutableDictionary<string, IPropertyPagesCatalog>> GetNamedCatalogsAsync()
+            {
+                if (catalogs != null)
                 {
-                    rule = configuredProjectExports.RuleFactory.CreateResolvedReferencePageRule(
-                                schema,
-                                context,
-                                dependency.Name,
-                                dependency.Properties);
+                    return catalogs.NamedCatalogs;
                 }
-                else
+
+                if (NamedCatalogs == null)
                 {
-                    rule = browseObjectsCatalog.BindToContext(schema.Name, context);
+                    // Note: it is unlikely that we end up here, however for cases when node providers
+                    // getting their node data not from Design time build events, we might have OnDependenciesChanged
+                    // event coming before initial design time build event updates NamedCatalogs in this class.
+                    // Thus, just in case, explicitly request it here (GetCatalogsAsync will acquire a project read lock)
+                    NamedCatalogs = await ActiveConfiguredProject.Services
+                        .PropertyPagesCatalog
+                        .GetCatalogsAsync(CancellationToken.None);
                 }
+
+                return NamedCatalogs;
             }
-            else
+        }
+
+        #endregion
+
+        #region Tree update cancellation
+
+        private CancellationTokenSource _treeUpdateCancellationTokenSource = new CancellationTokenSource();
+
+        private CancellationToken GetNextTreeUpdateCancellationToken()
+        {
+#pragma warning disable CA2000 // Dispose objects before losing scope
```

Comment:
Can you add a more descriptive comment about how we don't need to dispose this because `nextSource` becomes  `priorSource`?

Reply:
Done.